python/lsst/pipe/base/graph.py

```diff
@@ -48,25 +52,60 @@
 # ------------------------
 
 
-class QuantumGraphNodes:
-    """QuantumGraphNodes represents a bunch of nodes in an quantum graph.
+class QuantumIterData:
+    """Helper class for iterating over quanta in a graph.
+
+    `QuantumGraph.orderedQuanta` method needs to return topologically ordered
+    Quanta together with their dependencies. This class is used as a value
+    for iterator, it contains enumerated Quantum and its pre-requisites.
+
+    Parameters
+    ----------
+    quantumId : `int`
+        Index of this Quantum, unique but arbitrary integer.
+    quantum : `~lsst.daf.butler.Quantum`
+        Quantum corresponding to a graph node.
+    taksDef : `TaskDef`
```

Comment:
typo: `taksDef`

I'm also a bit more generally concerned that `TaskDef` is being used too broadly; it's conceptually just a `Task` class and ` Config` instance, and I think it's usually a good simplification of the design to flatten those out into separate attributes for other structs like this one and `QuantumGraphNodes`.

But maybe we should go further than that - `Quantum` actually already has its own `task` attribute.  That's not actually used anywhere (except to get/put the `Task` name from/to the database).  It doesn't have a `config` attribute, but perhaps it should (while allowing it to be `None`).  That would remove the need for any kind of task or config information here.

In fact, I think we should consider removing that information from `QuantumGraphNodes` as well, and then maybe flatten `QuantumGraph` into *just* a sequence of `Quantum`s.  That essentially denormalizes the task and config information into the individual `Quantums`, but I think that's a good thing: while a single preflight invocation will generate a `QuantumGraph` for which the current structure makes sense, if you instead obtain a `QuantumGraph` by querying provenance information from the database, the current structure does not make sense in general (if the provenance extends beyond a single run).  We haven't implemented that yet, but I certainly plan to.  I could also imagine us in the future wanting to run preflight multiple times with different configurations (or pipelines) for different data IDs, and merge the `QuantumGraph`s prior to execution.

Reply:
I agree that TsakDef looks somewhat artificial, and I think this deserves some discussion outside this ticket. I can add here that TaskDef also has `label` attribute and that serves as a sort of connection to original `Pipeline` from which graph was built.

Reply:
Good point about `label`.  Makes me wonder if we should add that to `Quantum`, too.  But as you said, we can discuss this on another ticket.